### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry/GammaSpecAdjunction): a missing lemma

### PR DESCRIPTION
**Pull request by jjaassoonn**
Added the following lemma:

Let $X$ be a locally ringed space, $R$ a ring and $U \subseteq \mathrm{Spec} R$ an open set.
Then for any $f : R \to \Gamma(\mathcal O_X, X)$, if we denote $F$ as the corresponding morphism $X \to \mathrm{Spec} R$ under the gamma spec adjunction  we have that 
the composition $res^{\mathrm{X}}\_{F^{-1}U} \circ f$ is equal to the composition $(R \to \mathcal{O}_{\mathrm{Spec} R}(U)\circ F(U)$
 
```lean
lemma toOpen_comp_locallyRingedSpaceAdjunction_homEquiv_app
    {X : LocallyRingedSpace} {R : Type u} [CommRing R]
    (f : LocallyRingedSpace.Γ.rightOp.obj X ⟶ op (CommRingCat.of R)) (U) :
    StructureSheaf.toOpen R U.unop ≫
      (locallyRingedSpaceAdjunction.homEquiv X (op <| CommRingCat.of R) f).1.c.app U =
    f.unop ≫ X.presheaf.map (homOfLE le_top).op := by
```

found usefuly by Andrew during #12371

Co-authored-by: Andrew Yang <the.erd.one@gmail.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean [hunk 1]
```diff
@@ -368,6 +368,33 @@ lemma locallyRingedSpaceAdjunction_counit :
     locallyRingedSpaceAdjunction.counit = (NatIso.op SpecΓIdentity.{u}).inv := rfl
 #align algebraic_geometry.Γ_Spec.LocallyRingedSpace_adjunction_counit AlgebraicGeometry.ΓSpec.locallyRingedSpaceAdjunction_counit
 
+lemma locallyRingedSpaceAdjunction_homEquiv_apply {X : LocallyRingedSpace} {R : CommRingCat}
+    (f : R ⟶ Γ.obj <| op X) :
+    locallyRingedSpaceAdjunction.homEquiv X (op R) (op f) =
+    identityToΓSpec.app X ≫ Spec.locallyRingedSpaceMap f := rfl
```
[the remark this answers is not in the document]
Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean]: done in [dfe2252](https://github.com/leanprover-community/mathlib4/pull/13412/commits/dfe2252282c9ce70b6727f626893e8e7f6fde8bf)

**Comment by jjaassoonn** [comment]
This is a fun error:
```bash
Error: ././././Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean:371:9: error: AlgebraicGeometry.ΓSpec.locallyRingedSpaceAdjunction_homEquiv_apply.{u_1} Left-hand side simplifies from
  (AlgebraicGeometry.ΓSpec.locallyRingedSpaceAdjunction.homEquiv X R) f
to
  (AlgebraicGeometry.ΓSpec.locallyRingedSpaceAdjunction.homEquiv X R) f
```

They look the same to me.

**Comment by erdOne** [comment]
Maybe either raise its priority or don't tag it as simp.

**Comment by jjaassoonn** [comment]
> Maybe either raise its priority or don't tag it as simp.

Raising priority to high doesn't seem to be working, so I deleted the simp attribute

File: Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean [hunk 2]
```diff
@@ -368,6 +368,34 @@ lemma locallyRingedSpaceAdjunction_counit :
...
+
+lemma toOpen_comp_locallyRingedSpaceAdjunction_homEquiv_app
+    {X : LocallyRingedSpace} {R : Type u} [CommRing R]
+    (f : LocallyRingedSpace.Γ.rightOp.obj X ⟶ op (CommRingCat.of R)) (U) :
+    StructureSheaf.toOpen R U.unop ≫
+      (locallyRingedSpaceAdjunction.homEquiv X (op <| CommRingCat.of R) f).1.c.app U =
+    f.unop ≫ X.presheaf.map (homOfLE le_top).op := by
+  rw [← StructureSheaf.toOpen_res _ _ _ (homOfLE le_top), Category.assoc]
+  show _ ≫ (Spec.locallyRingedSpaceObj (CommRingCat.of R)).presheaf.map
+    (homOfLE (le_top (a := U.unop))).op ≫ _ = _
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean:391]: I think this `show` is hiding a lot of missing (simp) lemmas. Can you add them?
Also the next `show`.

File: Mathlib/CategoryTheory/Opposites.lean [hunk 3]
```diff
@@ -246,6 +246,9 @@ protected def rightOp (F : Cᵒᵖ ⥤ D) : C ⥤ Dᵒᵖ where
   map f := (F.map f.op).op
 #align category_theory.functor.right_op CategoryTheory.Functor.rightOp
 
+lemma rightOp_map_unop {F : Cᵒᵖ ⥤ D} {X Y} (f : X ⟶ Y) :
+    (F.rightOp.map f).unop = (F.map f.op) := rfl
```
Comment (alreadydone) [inline Mathlib/CategoryTheory/Opposites.lean:250]:
```suggestion
    (F.rightOp.map f).unop = F.map f.op := rfl
```

File: Mathlib/AlgebraicGeometry/Spec.lean [hunk 4]
```diff
@@ -211,6 +211,26 @@ def Spec.locallyRingedSpaceObj (R : CommRingCat.{u}) : LocallyRingedSpace :=
...
+lemma Spec.locallyRingedSpaceObj_presheaf_map (R : CommRingCat.{u}) {U V} (i : U ⟶ V) :
+    (Spec.locallyRingedSpaceObj R).presheaf.map i =
+    (structureSheaf R).1.map i := rfl
+
+lemma Spec.locallyRingedSpaceObj_presheaf' (R : Type u) [CommRing R] :
+    (Spec.locallyRingedSpaceObj <| CommRingCat.of R).presheaf = (structureSheaf R).1 := rfl
+
+lemma Spec.locallyRingedSpaceObj_presheaf_map' (R : Type u) [CommRing R] {U V} (i : U ⟶ V) :
+    (Spec.locallyRingedSpaceObj <| CommRingCat.of R).presheaf.map i =
+    (structureSheaf R).1.map i := rfl
```
[the remark this answers is not in the document]
Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/Spec.lean:232]: `simps` generates the correct `presheaf` lemma, but the autogenerated `presheaf_map` lemma is not so good

File: Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean [hunk 5]
```diff
@@ -368,6 +368,72 @@ lemma locallyRingedSpaceAdjunction_counit :
...
+@[reassoc]
+lemma locallyRingedSpaceAdjunction_homEquiv_apply_naturality''
+    {X : LocallyRingedSpace} {R : Type u} [CommRing R]
+    (f : Γ.rightOp.obj X ⟶ op (CommRingCat.of R)) {U V} (i : U ⟶ V) :
+    (structureSheaf R).val.map i ≫
+      (locallyRingedSpaceAdjunction.homEquiv X (op <| .of R) f).1.c.app V =
+    (locallyRingedSpaceAdjunction.homEquiv X (op <| .of R) f).1.c.app U ≫
+      ((locallyRingedSpaceAdjunction.homEquiv X (op <| .of R) f).val.base _*
+        X.presheaf).map i :=
+  NatTrans.naturality _ _
```
[the remark this answers is not in the document]
Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean]: Thanks, these has been removed

File: Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean [hunk 6]
```diff
@@ -368,6 +368,72 @@ lemma locallyRingedSpaceAdjunction_counit :
     locallyRingedSpaceAdjunction.counit = (NatIso.op SpecΓIdentity.{u}).inv := rfl
 #align algebraic_geometry.Γ_Spec.LocallyRingedSpace_adjunction_counit AlgebraicGeometry.ΓSpec.locallyRingedSpaceAdjunction_counit
 
+lemma locallyRingedSpaceAdjunction_counit_app (R : CommRingCatᵒᵖ) :
+    (locallyRingedSpaceAdjunction.counit.app R).unop =
+    toOpen R.unop ⊤ := rfl
```
[the remark this answers is not in the document]
Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean]: done

File: Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean [hunk 7]
```diff
@@ -368,6 +368,43 @@ lemma locallyRingedSpaceAdjunction_counit :
     locallyRingedSpaceAdjunction.counit = (NatIso.op SpecΓIdentity.{u}).inv := rfl
 #align algebraic_geometry.Γ_Spec.LocallyRingedSpace_adjunction_counit AlgebraicGeometry.ΓSpec.locallyRingedSpaceAdjunction_counit
 
+@[simp]
+lemma locallyRingedSpaceAdjunction_counit_app (R : CommRingCatᵒᵖ) :
+    locallyRingedSpaceAdjunction.counit.app R =
+    (toOpen R.unop ⊤).op := rfl
```
Comment (joelriou) [inline Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean:374]: We have supposed to indent a bit after line breaks:
```suggestion
      (toOpen R.unop ⊤).op := rfl
```

Comment (joelriou) [reply Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean:374]: (Same below.)

File: Mathlib/CategoryTheory/Opposites.lean [hunk 8]
```diff
@@ -250,6 +250,9 @@ protected def rightOp (F : Cᵒᵖ ⥤ D) : C ⥤ Dᵒᵖ where
   map f := (F.map f.op).op
 #align category_theory.functor.right_op CategoryTheory.Functor.rightOp
 
+lemma rightOp_map_unop {F : Cᵒᵖ ⥤ D} {X Y} (f : X ⟶ Y) :
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Opposites.lean:253]:
```suggestion
@[simp]
lemma rightOp_map_unop {F : Cᵒᵖ ⥤ D} {X Y} (f : X ⟶ Y) :
```

Comment (erdOne) [reply Mathlib/CategoryTheory/Opposites.lean:253]: I think this is already provable by simp (`Functor.rightOp_map` + `Quiver.Hom.unop_op`)

Comment (jjaassoonn) [reply Mathlib/CategoryTheory/Opposites.lean:253]: indeed

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**